Refactor tests to use page object

alefox-pov opened the pull request:
- Refactor tests to use page object
- Rewrite DucotSearchForDocument test to check all documents within the table instead of looking for id by text
- Moved data to class fields, BaseTest class, and app.config

Commit:
Refactor tests to use page object

mikeformus commented on NDTraining/SeleniumTestTraining_O/Tests/DucotSaveSavedSearch.cs:
```diff
@@ -10,65 +11,31 @@
-            actions.PerformSimpleSearch(driver, wait, criteria);
-
-            driver.FindElement(By.Id("containerOptionsId")).Click();
+            objLogin = new LoginPage(driver);
```
Create separate `UI` class that accepts WebDriver in its constructor and holds references to all Page objects in the following way (so they are lazy-loaded):
```
private HomePage _homePage;
public HomePage HomePage => _homePage ?? (_homePage = new HomePage(driver));
```
Expose `UI` as a property in BaseTest.
It will allow you to access PageObjects in tests in the following way:
`UI.LoginPage.LoginToWebsite(userName, passWord);` without the need to create each page object in every test

mikeformus commented on NDTraining/SeleniumTestTraining_O/PageObjects/SearchResultsPage.cs:
```diff
@@ -0,0 +1,104 @@
+
+        public bool CheckForFoundDoc(string criteria)
+        {
+            bool result = false;
```
this method can be simplified if you use parameterized selector:
create method that returns selector for list view item by its name:
```
private By ListViewItem(string itemName) => By.Xpath($"//span[contains(@class,'lvName-span')][@title='{itemName}']");
```
then here:
```
WebDriverWait wait = new WebDriverWait(driver, TimeSpan.FromSeconds(15));
wait.Until(x => driver.FindElement(ListViewItem(criteria));
)
```
you can either make this method void - it will throw exception if item is not found - or wrap it into try catch and return true/false (and make assert statement in test)

mikeformus commented on NDTraining/SeleniumTestTraining_O/Tests/DucotSaveSavedSearch.cs:
```diff
@@ -10,65 +11,31 @@
-        public void CloseBrowser()
-        {
-            driver.Close();
+            Assert.IsTrue(objSearch.CheckSavedSearchName(savedSearchName));
```
you should use the same method as for verifying document in search results:
```
SearchResultsPage.CheckForFoundDoc(savedSearchName)
```
since when you save search results - next page is search results page with that saved seach in a list view.

mikeformus commented on NDTraining/SeleniumTestTraining_O/PageObjects/SearchResultsPage.cs:
```diff
@@ -0,0 +1,104 @@
+            driver.FindElement(searchResultsMenu).Click();
+        }
+
+        public void SelectFromMenuList(string selectOption)
```
can be simplified if you use Xpath with option name (see ListViewItem example)

mikeformus commented on NDTraining/SeleniumTestTraining_O/PageObjects/SearchResultsPage.cs:
```diff
@@ -0,0 +1,104 @@
+
+            try
+            {
+                return wait.Until(SeleniumExtras.WaitHelpers.ExpectedConditions.TextToBePresentInElementLocated(searchResultSavedName, searchName));
```
add `using SeleniumExtras.WaitHelpers;`

Commit:
- implemented lazy loading of Page Objects via Lazy class
- updated methods for selecting elements from menu and verifying found doc/saved search names

Commit:
removed unnecessary comment

mikeformus commented on NDTraining/SeleniumTestTraining_O/BaseTest.cs:
```diff
@@ -0,0 +1,37 @@
+            driver = new ChromeDriver(ConfigurationManager.AppSettings["ChromeDriver"]);
+            driver.Manage().Window.Maximize();
+            driver.Url = ConfigurationManager.AppSettings["BaseUrl"];
+            ui = new UI(driver);
```
`ui` and `driver` should be private. `ui` should be available in tests as property: `UI`

mikeformus commented on NDTraining/SeleniumTestTraining_O/PageObjects/CreateSavedSearchPage.cs:
```diff
@@ -0,0 +1,41 @@
+            this.driver = driver;
+        }
+
+        public void EnterSavedSearchName(string searchName)
```
return type `CreateSavedSearchPage`

mikeformus commented on NDTraining/SeleniumTestTraining_O/PageObjects/CreateSavedSearchPage.cs:
```diff
@@ -0,0 +1,41 @@
+            wait.Until(x => driver.FindElement(savedSearchName)).SendKeys(searchName);
+        }
+
+        public void ClickOKButton()
```
return type - `SearchResultsPage`

mikeformus commented on NDTraining/SeleniumTestTraining_O/PageObjects/CreateSavedSearchPage.cs:
```diff
@@ -0,0 +1,41 @@
+            driver.FindElement(savedSearchOKBtn).Click();
+        }
+
+        public void SaveSeach(string searchName)
```
return type - SearchResultsPage

mikeformus commented on NDTraining/SeleniumTestTraining_O/PageObjects/HomePage.cs:
```diff
@@ -0,0 +1,49 @@
+
+namespace SeleniumTestTraining_O
+{
+    class HomePage
```
Add return type instead of `void`

mikeformus commented on NDTraining/SeleniumTestTraining_O/PageObjects/LoginPage.cs:
```diff
@@ -0,0 +1,48 @@
+
+namespace SeleniumTestTraining_O
+{
+    class LoginPage
```
add return types instead of `void`

mikeformus commented on NDTraining/SeleniumTestTraining_O/PageObjects/SearchResultsPage.cs:
```diff
@@ -0,0 +1,64 @@
+        readonly IWebDriver driver;
+        readonly By searchResultsLoaded = By.Id("lvCheckMaster-input");
+        readonly By searchResultsMenu = By.Id("containerOptionsId");
+        readonly By menuOptionsItem = By.XPath($"//li[contains(@class,'saveSearch')]");
```
this should be parameterized to accept any option as a string

mikeformus commented on NDTraining/SeleniumTestTraining_O/PageObjects/SearchResultsPage.cs:
```diff
@@ -0,0 +1,64 @@
+            driver.FindElement(menuOptionsItem).Click();
+        }
+
+        public void SelectMenuOption()
```
this should accept option as a string

mikeformus commented on NDTraining/SeleniumTestTraining_O/UI.cs:
```diff
@@ -0,0 +1,26 @@
+            this.driver = driver;
+        }
+
+        public Lazy<LoginPage> LoginPage => new Lazy<LoginPage>(() => new LoginPage(driver));
```
No need to use Laze<> - it creates an overhead - we need to call Page.value everytime. Just use the approach that I described in my previous comment - when each property is backed by private field that gets initizlied only when accessed for the first time.

Commit:
- Replace lazy loading aproach
- Add return types instead of void in methods
- Add parameter when selecting options from search result menu
